firedrake/assemble.py

```diff
@@ -17,49 +17,61 @@
 from pyop2.exceptions import MapValueError, SparsityFormatError
 
 
-__all__ = ("assemble", )
+__all__ = ("assemble",)
 
 
-class AssemblyRank(IntEnum):
+class _AssemblyRank(IntEnum):
     SCALAR = 0
     VECTOR = 1
     MATRIX = 2
 
 
+class _AssemblyType(IntEnum):
+    SOLUTION = auto()
+    RESIDUAL = auto()
+
+
+_AssemblyOpts = namedtuple("_AssemblyOpts", ["diagonal",
+                                             "assembly_type",
+                                             "fc_params",
+                                             "mat_type",
+                                             "sub_mat_type",
+                                             "appctx",
+                                             "options_prefix"])
+"""Container to hold immutable assembly options."""
```

Comment:
I might be missing it, but it looks like the meaning of these options aren't documented anywhere?